src/Rules/Variables/CoalesceRule.php

```diff
@@ -0,0 +1,74 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Rules\Variables;
+
+use PhpParser\Node;
+use PHPStan\Analyser\Scope;
+use PHPStan\Rules\RuleError;
+use PHPStan\Rules\RuleErrorBuilder;
+use PHPStan\Type\NullType;
+use PHPStan\Type\VerbosityLevel;
+
+/**
+ * @implements \PHPStan\Rules\Rule<\PhpParser\Node\Expr\BinaryOp\Coalesce>
+ */
+class CoalesceRule implements \PHPStan\Rules\Rule
+{
+
+	public function getNodeType(): string
+	{
+		return Node\Expr\BinaryOp\Coalesce::class;
+	}
+
+	public function processNode(Node $node, Scope $scope): array
+	{
+		$error = $this->canBeCoalesced($node->left, $scope);
+
+		if ($error === null) {
+			return [];
+		}
+
+		return [$error];
+	}
+
+	private function canBeCoalesced(Node $node, Scope $scope): ?RuleError
+	{
+		if ($node instanceof Node\Expr\Variable && is_string($node->name)) {
+
+			$hasVariable = $scope->hasVariableType($node->name);
+
+			if ($hasVariable->no()) {
+				return RuleErrorBuilder::message(
+					sprintf('Coalesce of undefined variable $%s.', $node->name)
+				)->line($node->getLine())->build();
+			}
+
+			$variableType = $scope->getVariableType($node->name);
+
+			if ($variableType->isSuperTypeOf(new NullType())->no()) {
```

Comment:
In this place the variable might not exist ($hasVariable might be maybe()) so usage with non-nullable variable is valid.

Reply:
We also need to check for variable that have `$hasVariable->yes()` and are not null. In that case left side is always used.